source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1632 @@
+                                                             const std::set<types::boundary_id> &boundary_ids) const
+    {
+      if (this->get_timestep_number() == 0)
+        return;
+
+      TimerOutput::Scope timer_section(this->get_computing_timer(), "FastScape plugin");
+      const int current_timestep = this->get_timestep_number ();
+      const double a_dt = this->get_timestep()/year_in_seconds;
+
+      //Vector to hold the velocities that represent the change to the surface.
+      std::vector<double> V(array_size);
+
+      // FastScape requires multiple specially defined and ordered variables sent to its functions. To make
+      // the transfer of these down to one process easier, we first fill out a vector of temporary variables,
+      // then when we get down to one process we use these temporary variables to fill the double arrays
+      // in the order needed for FastScape.
+      std::vector<std::vector<double>> temporary_variables = get_aspect_values();
+
+      // Run FastScape on single process.
+      if (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0)
+        {
+          // Initialize the variables that will be sent to FastScape.
+          // These have to be doubles of array_size, which C++ doesn't like,
+          // so they're initialized this way.
+          std::unique_ptr<double[]> h (new double[array_size]());
+          std::unique_ptr<double[]> vx (new double[array_size]());
+          std::unique_ptr<double[]> vy (new double[array_size]());
+          std::unique_ptr<double[]> vz (new double[array_size]());
+          std::unique_ptr<double[]> kf (new double[array_size]());
+          std::unique_ptr<double[]> kd (new double[array_size]());
+          std::unique_ptr<double[]> b (new double[array_size]());
+          std::unique_ptr<double[]> sf (new double[array_size]());
+          std::vector<double> h_old(array_size);
+
+          fill_fastscape_arrays(h.get(), kd.get(), kf.get(), vx.get(), vy.get(), vz.get(), temporary_variables);
+
+          if (current_timestep == 1 || restart)
+            {
+              this->get_pcout() << "   Initializing FastScape... " << (1+maximum_surface_refinement_level+additional_refinement) <<
+                                " levels, cell size: " << dx << " m." << std::endl;
+
+              // If we are restarting from a checkpoint, load h values for FastScape instead of using the ASPECT values.
+              if (restart)
+                {
+                  read_restart_files(h.get(), b.get(), sf.get());
+
+                  restart = false;
+                }
+
+              initialize_fastscape(h.get(), b.get(), kd.get(), kf.get(), sf.get());
+            }
+          else
+            {
+              // If it isn't the first timestep we ignore initialization and instead copy all height values from FastScape.
+              if (use_velocities)
```

Comment:
This was added in case someone wanted to use the free surface in combination with fastscape, in which case fastscape won't advect/uplift the mesh. Generally, this is set to true and we copy the likely higher resolution surface topography directly from fastscape. Otherwise, if the free surface is on, we wanted to use the updated/advected surface we gathered from ASPECT.